[SPARK-16440][MLlib] Ensure broadcasted variables are destroyed even in case of exception

AnthonyTruchet opened the pull request:
## What changes were proposed in this pull request?

Ensure broadcasted variable are destroyed even in case of exception
## How was this patch tested?

Word2VecSuite was run locally

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala:
```diff
@@ -313,133 +313,139 @@ class Word2Vec extends Serializable with Logging {
+      val wordArray = vocab.map(_.word)
+      new Word2VecModel(wordArray.zipWithIndex.toMap, syn0Global)
+    }
+    finally
```
Not sure this will pass the style checker. My only hesitation is that really if we do this one place we should do it in 100 places, and the exception path here is not a usual one.

AnthonyTruchet replied:
btw. are you aare of http://jsuereth.com/scala-arm/ which might prove damn useful to ease resource management without clustering (much) the syntax. It is akin to using clauses in C# or context managers in Python.

pzz2011 commented:
@AnthonyTruchet 　If not Ensure broadcasted variables destoryed, what will happend?(this is my person confusion:)

AnthonyTruchet commented:
@pzz2011 Nothing _that_ bad if your application is just learning once and saving / using the embeddings. But if you run hundreds or thousands of learning, you'll just crash because of uncontrolled memory usage... and that's bad !

thunterdb commented:
@AnthonyTruchet thank you for the PR. This is definitely worth fixing for large deployments. Now, as you noticed, this portion of code does not quite abide by the best engineering practices... Instead of adding an extra layer of nesting, would you mind make the following changes?

```scala
  def fit[S <: Iterable[String]](dataset: RDD[S]): Word2VecModel = {
    ...
    val expTable = sc.broadcast(createExpTable())
    val bcVocab = sc.broadcast(vocab)
    val bcVocabHash = sc.broadcast(vocabHash)
    try { fit0(expTable, ...) } finally {
      ...
    }
   }
   private final def fit0(...) { 
     // Put all the content here.
     // Note that the inner code also includes some broadcasts, you may want to fix these as well if you can
   }
```

I personally agree about resource management and scala-arm. We try to keep scala dependencies to a minimum, unfortunately, because they can be very tedious to move from one scala version to another.

AnthonyTruchet commented:
@thunterdb Copy that, working on it and sorry for the acknowledge delay.

vanzin commented:
You also should file a bug and reference it from the PR title.

Commit:
[SPARK-16440][MLlib] Destroy broadcasted variables in a try finally

Extracted the do_fit method in order to keep nesting under control.

This contribution is on done on behalf of Criteo (http://labs.criteo.com/),
according to the terms of the Apache 2.0 license.

AnthonyTruchet commented:
@vanzin The ticket was filled a long time ago, I updated the PR to make it clearer. Is any manual linking in some other forge needed ?

@thunterdb I have updated the review following your suggestion, without using ARM: ready fo review.

I'll check with the team working on W2V how to could contribute to  make this piece of code "follow better engineering practice" :-)

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala:
```diff
@@ -314,6 +315,20 @@ class Word2Vec extends Serializable with Logging {
     val expTable = sc.broadcast(createExpTable())
     val bcVocab = sc.broadcast(vocab)
     val bcVocabHash = sc.broadcast(vocabHash)
+      do_fit(dataset, sc, expTable, bcVocab, bcVocabHash)
```
Are we missing a try? The formatting should be

```
try {
  ...
} finally {
  ...
}
```

Also use camelCase rather than underscore_naming

AnthonyTruchet replied:
my bad, local rebase issue.

vanzin commented:
ok to test

SparkQA commented:
**[Test build #73600 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73600/testReport)** for PR 14299 at commit [`e716700`](https://github.com/apache/spark/commit/e716700a5b684d6a985860ea164996471a7341ef).

SparkQA commented:
**[Test build #73600 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73600/testReport)** for PR 14299 at commit [`e716700`](https://github.com/apache/spark/commit/e716700a5b684d6a985860ea164996471a7341ef).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73600/
Test FAILed.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala:
```diff
@@ -314,6 +315,21 @@ class Word2Vec extends Serializable with Logging {
+    try {
+      doFit(dataset, sc, expTable, bcVocab, bcVocabHash)
+    }
+    finally
```
Put this on one line but otherwise seems reasonable. The broadcast is unused after fitting. Are there more instances like this?

AnthonyTruchet replied:
I already spotted and corrected a few instances of forgotten broadcasted variables. I've already reported and patched all I was aware off.

I was suggesting introducing an ARM pattern and making broadcasted variable ARM aware (see comment above). This is by no way a general solution but it might help reduce the issue when the use of a broadcasted variable is compatibble with a syntactic scope. @thunterdb legitimately raised concern about introducing new dependencies (to scala-arm specifically). What is your insight wrt to this ?

AnthonyTruchet replied:
Formatting and scala style check fixed.

SparkQA commented:
**[Test build #73685 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73685/testReport)** for PR 14299 at commit [`9a0675f`](https://github.com/apache/spark/commit/9a0675f4fbf34d2095710628fb5ca3adf98f7f5b).

SparkQA commented:
**[Test build #73685 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73685/testReport)** for PR 14299 at commit [`9a0675f`](https://github.com/apache/spark/commit/9a0675f4fbf34d2095710628fb5ca3adf98f7f5b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73685/
Test FAILed.

srowen reviewed (changes requested):
This still doesn't match the Spark code style at all. Have a look at how other try blocks and method params are formatted and match it.

SparkQA commented:
**[Test build #73699 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73699/testReport)** for PR 14299 at commit [`6b8ae85`](https://github.com/apache/spark/commit/6b8ae85dc362ebef0f8d416a8e35970f57130a9f).

SparkQA commented:
**[Test build #73699 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73699/testReport)** for PR 14299 at commit [`6b8ae85`](https://github.com/apache/spark/commit/6b8ae85dc362ebef0f8d416a8e35970f57130a9f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73699/
Test FAILed.

SparkQA commented:
**[Test build #3588 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3588/testReport)** for PR 14299 at commit [`6b8ae85`](https://github.com/apache/spark/commit/6b8ae85dc362ebef0f8d416a8e35970f57130a9f).

SparkQA commented:
**[Test build #3588 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3588/testReport)** for PR 14299 at commit [`6b8ae85`](https://github.com/apache/spark/commit/6b8ae85dc362ebef0f8d416a8e35970f57130a9f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

srowen commented:
Oh, these broadcasts are already destroyed actually, inside the method. This isn't needed then. I kinda thought we had taken care of most or all of these already.

AnthonyTruchet commented:
I'm fixing this (issue compiling Spark locally delay me). The whole point is that they are *not* destroyed within the method in case of exception.

srowen commented:
Yes, but they are still cleaned up (eventually). It can only happen one place. Unless the exception path is common I don't know if it's worth changing this everywhere, because for consistency it would really apply everywhere, and we declined to do that before.

AnthonyTruchet commented:
Actually we *did* observe memory leak (severe enough to lead to application failure) due to this when doing thousands of Word2Vec learning in one Spark application, a few of them failing.

srowen commented:
I wonder how that happens and why that fixes this though. It would be the same before and after except for the blocking flag. That should be false like all the other calls ... Unless that is somehow part of the issue?

AnthonyTruchet commented:
The blocking flag is not the issue.
The issue is that there was not try / finally and the destruction was only done if the fitting succeded and skipped otherwise.

In this PR I move the destruction into a try / finally construct, all the while extracting the actual fitting into a dedicated method to keep nesting under control.

Whether the destruction should be blocking or not is pretty debatable and I do not have any strong mind on the question, I choose to make them blocking "just to be sure" this is actually performed. If you'd rather keep the non blocking best effort current behaviour I'll perform the change.

SparkQA commented:
**[Test build #73770 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73770/testReport)** for PR 14299 at commit [`65a238d`](https://github.com/apache/spark/commit/65a238d1cd09c96b41337d3d8f5162fca9faf2f4).

SparkQA commented:
**[Test build #73770 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73770/testReport)** for PR 14299 at commit [`65a238d`](https://github.com/apache/spark/commit/65a238d1cd09c96b41337d3d8f5162fca9faf2f4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

srowen commented:
I am still not getting why the failure of this method is a normal execution path - when would it fail repeatedly?

Yes the destroy should be non-blocking for consistency.

AnthonyTruchet commented:
For the "normal failure case", the one we encounter is that when the vocabulary is too small the embedding learning raises. And it is pretty legitimate in our use cases to have some "too small vocabularies".

srowen commented:
OK, if this happens normally enough that you would accumulate a lot of broadcasts that aren't going out of scope and hogging resources, I can see special-casing this with try-finally. I'm still surprised it's so common so quickly but am OK with it -- worst case it just doesn't hurt consistency much to worry about.

AnthonyTruchet commented:
Ok, thanks. Do you  sen any missing task for it to be merged then ?

srowen commented:
Merged to master
